### DISCUSSION TITLE
📝【LeetCode】84 柱状图中最大的矩形

### DISCUSSION THREAD

**Discussion by imageslr** [imageslr/blog#77] [Announcements]
https://imageslr.github.io/2020/02/02/leetcode-84.html 

 这道题是 LeetCode 84 题。

**Comment by zyksir** [comment]
优化解法4要怎么做到 O(n)呢？我想到的代码是这样子的
```python3
j = i - 1
while height[j] > height[i]:
    j = dp[j]
```

**Comment by imageslr** [comment]
@zyksir 因为我们要找到一个位置左、右**第一个比它矮**的柱子，所以需要从左往右、从右往左分别算一次。你这个公式是从左往右算，再结合一个从右往左算就可以了。

可以看看 [42 题 - 接雨水的解法四](https://imageslr.com/2020/02/21/leetcode-42.html#%E8%A7%A3%E6%B3%95%E5%9B%9B%E4%BD%BF%E7%94%A8%E5%8A%A8%E6%80%81%E8%A7%84%E5%88%92%E4%BC%98%E5%8C%96%E8%A7%A3%E6%B3%95%E4%BA%8C)，就是从左往右、从右往左算了两次左边、右边**最高**的柱子。
